### ISSUE TITLE
error handling for unmarshalling

### ISSUE THREAD

**Issue by nikolaydubina** [zarldev/goenums#7]
this creates new value and overwrites by value-copy struct stored at pointer
```go
func (t *Status) UnmarshalJSON(b []byte) error {
	*t = Parse(string(b))
	return nil
}
```

this is concerning:
1. ~does this even work? usually if you have struct at pointer, memory is already allocated by Unmarshal package. here normally you would override internal field instead.~ actually it is ok to override whole struct. it is small anyways. very convenient too, as it works for any types.
2. ~allocating enum size twice. Parse creates new enum json.Unmarshal creates enum. you can save space if you move Parse into this this code.~ for small enums like `struct{uint8}` it is ok to allocate twice.
3. ~coverting bytes to string may cause overhead, since string is immutable, and bytes are mutable. Go would have to make copy of data stored at bytes. access bytes directly. Benchmakrs should show that.~ efficient map lookups need string. 
4. lastly, clearly sometimes you can't unmarshal enum. would be much better if we did not silently fail (which is root of so many problems). would be good if we error out if value is not valid. (or control this via gencode flag)

**Comment by zarldev** [comment]
Yeah - tbh I like the idea of failing on an invalid enum parse - definitely should be a generation flag.

**Comment by matdurand** [comment]
I would also like to have that. Fallbacking on the `invalid` value is probably nice in some cases, but in my case I want the unmarshalling to fail. 

In general, I understand the need for the invalid value to account for the "0" value, but in practice it feels a bit alien to work with. Instead of a generator flag, could we generate 2 types of enums? The first one with a possible invalid value, which doesn't fail on parsing and fallback to the invalid value, and a second one without an invalid value that is more strict and that would fail on parsing/unmarshalling.

```
type discountType int
//no invalid value, generates a strict enum
const (
	percentage discountType = iota + 1
	amount
	giveaway
)
```
(I'm getting a panic when trying this btw)

**Comment by zarldev** [comment]
@matdurand this was failing due to the non 0 indexed enum - added to test cases and have release v0.2.7 with the fix.

**Comment by zarldev** [comment]
This has been added @matdurand there is now the `-f`/`-failfast`  flag for enabling error on Json unmarshall to a Invalid enum.

**Comment by zarldev** [comment]
> I would also like to have that. Fallbacking on the `invalid` value is probably nice in some cases, but in my case I want the unmarshalling to fail.
> 
> In general, I understand the need for the invalid value to account for the "0" value, but in practice it feels a bit alien to work with. Instead of a generator flag, could we generate 2 types of enums? The first one with a possible invalid value, which doesn't fail on parsing and fallback to the invalid value, and a second one without an invalid value that is more strict and that would fail on parsing/unmarshalling.

You could use goenums to generate your 2 enum types and with embedding and the Unmarshal interface have a wrapper struct

```golang

type Combo struct {
	EnumWithInvalid
	StrictEnum
}

func (c *Combo) Unmarshal(b []byte) error {
	var e EnumWithInvalid
	err := e.Unmarshal(b)
	if err != nil {
		return err
	}
	c.EnumWithInvalid = e
	var s StrictEnum
	err = s.Unmarshal(b)
	if err != nil {
		return err
	}
	return nil
}
```
Does that work for your use case? This will auto update when the enums are regenerated so it is a one of cost..

**Comment by matdurand** [comment]
Not sure why I would need this Combo wrapper to be honest. Maybe I didn't explain myself enough. My original comment was about 2 things:

1. Making sure that the JSON unmarshalling could fail on invalid value, which is now taken care of by the new failfast flag.
2. Making sure any other potential call to `ParseXXX` would also fail for non JSON case, which is still a problem.

Why not adjust the failfast to make sure `ParseXXX` returns a value + potential error? Since `UnmarshalJSON` calls `ParseXXX`, it would kill 2 birds with one stone ... But that would mean changing the signature of the Parse method to now return an value+error.

Basically my main point was that in most cases I don't think people would care much about the "invalid" value and would rather receive an error instead of a value which they then need to compare to see if it's valid or not. Using the error pattern seems more golang idiomatic in that regard.

**Comment by zarldev** [comment]
@matdurand @nikolaydubina 
See changes made in v0.3.0 I have modified the ParseXXX methods to return an error.

**Comment by matdurand** [comment]
Amazing! Thank you
